### ISSUE TITLE
How to reproduce the result of the Metric-based Optimization model mentioned in the Paper?

### ISSUE THREAD

**Issue by BlueZeros** [yafuly/MgMO-NAT#1]
## What is your question?
We fail to reproduce the result of the Metric-based Optimization model baseline mentioned in the Paper. The code repository only contains the implement of the MgMO method. 

## What have you tried
Following the setting in the paper, we use metirc-based object to fine-tune the pre-trained CMLM. We pass the decoder only once to get the output distribution during training process. Then we use the function `torch.multinomial()` to sample 10 sentences from the output distribution to calculate the reward. 

The parameters we use are shown below:
* learning rate = 5e-6
* lr scheduler =  inverse_sqrt
* dropout rate = 0.3
* warmup iteration = 5000
* fintune iteration = 30000
 
We average 5 best checkpoints on validation set and the test command 
```
CUDA_VISIBLE_DEVICES=0 fairseq-generate \
    $DATA  \
    --gen-subset test \
    --noise full_mask \
    --task translation_lev \
    --path $SAVE/averaged_model.pt \
    --iter-decode-with-beam 5 \
    --iter-decode-max-iter 0 \
    --iter-decode-force-max-iter \
    --iter-decode-eos-penalty 0 \
    --beam 1 --remove-bpe \
    --print-step \
    --batch-size 50 | tee $SAVE/generate.out
```

The final result on WMT14 EN-DE distilled dataset  is 
`Generate test with beam=1: BLEU4 = 21.61, 50.8/27.1/16.1/9.9 (BP=1.000, ratio=1.151, syslen=74216, reflen=64506)`
which is lower than the result 24.8 shown in the paper.

I would appreciate it if you can release the implement of the Metric-based Optimization model baseline!

**Comment by yafuly** [comment]
Hi!

Thanks for your interest in our work.

Based on what you have done, there are two ways to proceed:

(1) To reproduce the Metric-based Optimization baseline (MO baseline), you can still use the script _run.sh_. But you need to modify line 175 in _/models/nat/cmlm_transformer.py_ to let the decoder receive all '\<unk\>' tokens and predict all reference tokens. A fast implementation: replace the line 174 and 175 with 
` prev_output_tokens = prev_output_tokens.repeat_interleave(S,0)`
The original code contains implementation for the MO baseline, but I removed it for simplicity :)

(2) Another way, following your script, there are three suggestions:
a. increase the sample size from 10 to 40.
b. choose the best validation checkpoint (it's usually better than the averaged one from my empirical results).
c. remove duplicated tokens in the hypothesis, as also suggested by previous work.

Contact me if you have further questions!

**Comment by yafuly** [comment]
I'm closing this issue as there are no further questions. But feel free to open another issue if there are any.

**Comment by yafuly** [comment]
I'm closing this issue as there are no further questions. But feel free to open another issue if there are any.
